fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java

```diff
@@ -183,8 +193,65 @@ private void prepare(ConnectContext connectContext, OptExpression logicOperatorT
         context.setTraceInfo(new OptimizerTraceInfo(connectContext.getQueryId()));
 
         // process materialized views
-        if (Config.enable_experimental_mv) {
+        // TODO: add session variable
+        if (Config.enable_experimental_mv && !optimizerConfig.isRuleBased()) {
             // TODO: register materialized views
+            // register materialized views
+            registerMaterializedViews(logicOperatorTree, connectContext);
+        }
+    }
+
+    private void registerMaterializedViews(OptExpression logicOperatorTree, ConnectContext connectContext) {
+        List<Table> tables = Utils.getAllTables(logicOperatorTree);
+
+        // get all related materialized views, include nested mvs
+        Set<MaterializedView> relatedMvs =
+                Utils.getRelatedMvs(connectContext.getSessionVariable().getNestedMvRewriteMaxLevel(), tables);
+
+        for (MaterializedView mv : relatedMvs) {
+            if (!mv.isActive()) {
+                continue;
+            }
+            Set<String> partitionNamesToRefresh = mv.getPartitionNamesToRefreshForMv();
+            PartitionInfo partitionInfo = mv.getPartitionInfo();
+            if (partitionInfo instanceof SinglePartitionInfo) {
+                if (!partitionNamesToRefresh.isEmpty()) {
+                    continue;
+                }
+            } else if (partitionNamesToRefresh.containsAll(mv.getPartitionNames())) {
+                // if the mv is partitioned, and all partitions need refresh,
+                // then it can not be an candidate
+                continue;
+            }
+
+            // 1. build mv query logical plan
+            ColumnRefFactory columnRefFactory = new ColumnRefFactory();
+            MaterializedViewOptimizer mvOptimizer = new MaterializedViewOptimizer();
+            OptExpression mvPlan = mvOptimizer.optimize(mv, columnRefFactory, connectContext);
```

Comment:
I think needn't generate mv plan every time, we should store it

Reply:
the reasons do not cache the plan are:
1. maybe some SessionVariable will change the plan generated
2. during each optimization, I will modify the mv plan to add some predicate to plan